apps/vanilla/src/pages/login/login.ts

```diff
@@ -0,0 +1,71 @@
+import M from 'materialize-css';
+
+import { UserLoginForm } from '@js-camp/core/forms/userLoginForm';
+
+import { AuthService } from '../../api/services/authService';
+import {
+  validatePassword,
+  validatePasswordField,
+} from '../../../utils/passwordValidation';
+import { AppUrl } from '../../../utils/constants';
+import { ApiError } from '../../api/errors';
+import {
+  validateEmail,
+  validateEmailField,
+} from '../../../utils/emailValidation';
+import { processFormChanged } from '../../../utils/formProcessing';
+import { getElementOrRaiseError } from '../../../utils/query';
+
+document.addEventListener('DOMContentLoaded', () => {
+  M.AutoInit();
+  const loginForm = getElementOrRaiseError<HTMLFormElement>('.login-form');
+  const submitButton = getElementOrRaiseError<HTMLButtonElement>('#login-submit-button');
+  const baseErrorBlock = getElementOrRaiseError<HTMLElement>('.base-error__text');
+  const passwordField = getElementOrRaiseError<HTMLInputElement>('#password');
+  const emailField = getElementOrRaiseError<HTMLInputElement>('#email');
+
+  loginForm.addEventListener('input', () => {
+    processFormChanged(
+      baseErrorBlock,
+      submitButton,
+      () => !validateEmail(emailField.value) || !validatePassword(passwordField.value),
+    );
+  });
+  loginForm.addEventListener('submit', processSubmitForm);
+  passwordField.addEventListener('input', () => validatePasswordField(passwordField));
+  emailField.addEventListener('input', () => validateEmailField(emailField));
+});
+
+/**
+ * Send request to login user.
+ * @param loginForm User login form.
+ */
+async function loginUser(loginForm: UserLoginForm): Promise<void> {
+  const baseErrorBlock = getElementOrRaiseError<HTMLElement>('.base-error__text');
+  try {
+    await AuthService.login(loginForm);
+    window.location.href = AppUrl.Base;
+  } catch (error: unknown) {
+    if (error instanceof ApiError) {
+      baseErrorBlock.innerText = error.message;
+    } else {
+      baseErrorBlock.innerText = 'Something went wrong';
+    }
+  }
+}
+
+/**
+ * Process form on submit.
+ * @param event Submit event.
+ */
+function processSubmitForm(event: SubmitEvent): void {
+  event.preventDefault();
+  const passwordField = getElementOrRaiseError<HTMLInputElement>('#password');
+  const emailField = getElementOrRaiseError<HTMLInputElement>('#email');
+
+  if (!validatePassword(passwordField.value) || !validateEmail(emailField.value)) {
+    return;
+  }
```

Comment:
Please, try not to use `!` when it's possible, in your case you could use De Morgan's law:
```suggestion
  if (validatePassword(passwordField.value) && validateEmail(emailField.value)) {
    return;
  }
```

Reply:
And it seems that when all validations are correct, nothing happens. Is it intended?